File: 2021/10/14/increasing_bst.py
"""
https://leetcode.com/problems/increasing-order-search-tree/
Author: Andrew Jarombek
Date: 10/14/2021
"""

from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def increasing_bst(self, root: TreeNode) -> Optional[TreeNode]:
        self.new_tree_start = None
        self.new_tree_end = None

        self.dfs(root)

        return self.new_tree_start

    def dfs(self, root: TreeNode) -> None:
        if root.left is not None:
            self.dfs(root.left)

        if self.new_tree_start is None:
            self.new_tree_start = TreeNode(root.val)
            self.new_tree_end = self.new_tree_start
        else:
            self.new_tree_end.right = TreeNode(root.val)
            self.new_tree_end = self.new_tree_end.right

        if root.right is not None:
            self.dfs(root.right)

File: 2023/05/11/increasingBST.js
/**
 * https://leetcode.com/problems/increasing-order-search-tree/
 * @author Andrew Jarombek
 * @since 5/11/2023
 */

const assert = require('assert');

function TreeNode(val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
}

const increasingBST = function (root) {
    const result = new TreeNode(-1);
    dfs(root, result);
    return result.right;
}

const dfs = function (node, tail) {
    if (node.left) {
        tail = dfs(node.left, tail);
    }

    tail.right = new TreeNode(node.val);
    tail = tail.right;

    if (node.right) {
        tail = dfs(node.right, tail);
    }

    return tail;
}

// Test case 1: Test the conversion of a small binary tree to an increasing order BST
// Create the binary tree
let root = new TreeNode(2);
root.left = new TreeNode(1);
root.right = new TreeNode(3);

// Perform the conversion
let result = increasingBST(root);

// Assert the expected output
assert.strictEqual(result.val, 1);
assert.strictEqual(result.right.val, 2);
assert.strictEqual(result.right.right.val, 3);
assert.strictEqual(result.right.right.right, null);

// Test case 2: Test the conversion of a single-node binary tree
// Create a single-node binary tree
root = new TreeNode(1);

// Perform the conversion
result = increasingBST(root);

// Assert the expected output
assert.strictEqual(result.val, 1);
assert.strictEqual(result.right, null);
